STA130 HOMEWORK ONE - EVELYN JIN

PART ONE: PRELECTURE

In [14]:
import pandas as pd
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv('https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv')
df.isna().sum()

row_n           0
id              1
name            0
gender          0
species         0
birthday        0
personality     0
song           11
phrase          0
full_id         0
url             0
dtype: int64

PART TWO: PRELECTURE

2.1

In [11]:
# Check the number of rows and columns
num_rows, num_columns = df.shape
print(f"Number of rows (observations): {num_rows}")
print(f"Number of columns (variables): {num_columns}")

# Optionally, print column names
print(f"Column names: {df.columns.tolist()}")

Number of rows (observations): 391
Number of columns (variables): 11
Column names: ['row_n', 'id', 'name', 'gender', 'species', 'birthday', 'personality', 'song', 'phrase', 'full_id', 'url']


2.2

Observations (Rows):
In the context of the dataset, an observation corresponds to a unique villager. Each row represents an individual character.

Variables (Columns):
A variable is an attribute recorded for each observation (villager). Each column in the DataFrame represents a distinct characteristic such as their name, birthday, species, or personality type. These variables describe the individual traits of each villager.

PART THREE: PRELECTURE

In [15]:
import pandas as pd

# Load the dataset (replace 'villagers.csv' with the correct path)
df = pd.read_csv('https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv')

# 1. Display basic summary statistics for numeric columns
print("Summary Statistics for Numeric Columns:")
print(df.describe())

# 2. Show unique value counts for categorical columns
# Example for the 'species' column (you can replace 'species' with other categorical columns)
print("\nValue Counts for 'species' Column:")
print(df['species'].value_counts())

# 3. If you want summaries for more categorical columns, do the same for other columns:
# Value counts for 'personality'
print("\nValue Counts for 'personality' Column:")
print(df['personality'].value_counts())

Summary Statistics for Numeric Columns:
            row_n
count  391.000000
mean   239.902813
std    140.702672
min      2.000000
25%    117.500000
50%    240.000000
75%    363.500000
max    483.000000

Value Counts for 'species' Column:
species
cat          23
rabbit       20
frog         18
squirrel     18
duck         17
dog          16
cub          16
pig          15
bear         15
mouse        15
horse        15
bird         13
penguin      13
sheep        13
elephant     11
wolf         11
ostrich      10
deer         10
eagle         9
gorilla       9
chicken       9
koala         9
goat          8
hamster       8
kangaroo      8
monkey        8
anteater      7
hippo         7
tiger         7
alligator     7
lion          7
bull          6
rhino         6
cow           4
octopus       3
Name: count, dtype: int64

Value Counts for 'personality' Column:
personality
lazy      60
normal    59
cranky    55
snooty    55
jock      55
peppy     49
smug      34
uchi      24
Name: count,

The .describe() method provides summary statistics for numeric columns in your dataset, such as the count, mean, standard deviation, minimum, and maximum values.

The .value_counts() method gives you a count of unique values for categorical columns. For example, .value_counts(personality) will show how many villagers belong to each personality. It can work for any other categorical column.

PART FOUR: PRELECTURE

In [16]:
import pandas as pd

# Load the Titanic dataset
df = pd.read_csv("https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv")

# Check the shape of the dataset
print(df.shape)

# Check the basic summary statistics
print(df.describe())

(891, 15)
         survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200


The df.shape function returns a tuple (number_of_rows, number_of_columns). This tells you the total size of the dataset. The output tells you that there are 891 rows (observations) and 15 columns (variables).

a) Number of Columns Analyzed

df.shape reports 15 columns (the total number of variables). df.describe() reports only the numeric columns, typically fewer than 15. df.describe() automatically excludes non-numeric columns (like sex, embarked, who, deck, etc.). in this dataset, only numeric columns like survived, pclass, age, sibsp, parch, and fare are included in the summary statistics.

(b) Values Reported in the "count" Column:

The "count" column in df.describe() reports how many non-missing values (i.e., non-NaN values) are present for each numeric column. Discrepancy in "count" values occur since missing values cause discrepancies. For example, in the age column, df.shape shows that the dataset has 891 rows, but the count for age is 714. This means there are 177 missing values (NaN) in the age column. The age column has a count of 714, meaning there are 714 non-missing values and 177 missing (NaN) values. Other columns like pclass, survived, and fare have no missing values and show a count of 891.

The differences occur because non-numeric columns such as df.describe() only analyzes numeric data by default, so columns like sex, embarked, deck, etc., which are categorical, are excluded from the analysis. It also happens because of missing values like the "count" in df.describe() reflects the number of non-missing values for each column. If there are missing values (like in the age column), the count will be lower than the total number of rows (which df.shape gives you).

PART FIVE: PRELECTURE

Attributes vs. Methods

Attributes are properties associated with an object. They do not have parentheses because they are simply values or states of the object. You access attributes directly, like df.shape to get the shape of the DataFrame.

Methods are functions that belong to an object. They perform actions on the object and are called using parentheses. You call methods to perform operations, like df.describe() to get a statistical summary of numeric data.

PART SIX: POSTLECTURE

These statistics provide a comprehensive summary of the distribution and spread of the numeric data in each column. Here are brief definitions of each.

Count: The number of non-missing values in the column. It indicates how many values were present for that particular variable.

Mean: The average value of the column. It is calculated by summing all the values in the column and then dividing by the count of non-missing values.

Std (Standard Deviation): A measure of the amount of variation or dispersion in the column's values. A high standard deviation indicates that values are spread out over a wider range, while a low standard deviation indicates that values are closer to the mean.

Min (Minimum): The smallest value in the column. It represents the lowest boundary of the data range.

25% (25th Percentile or First Quartile): The value below which 25% of the data falls. It’s a measure of the lower boundary of the middle 50% of the data.

50% (50th Percentile or Median): The middle value of the column when the data is sorted in ascending order. It divides the data into two equal halves.

75% (75th Percentile or Third Quartile): The value below which 75% of the data falls. It’s a measure of the upper boundary of the middle 50% of the data.

Max (Maximum): The largest value in the column. It represents the upper boundary of the data range.

PART SEVEN: POSTLECTURE

The Titanic dataset typically contains columns like:

pclass: Passenger class

sex: Gender

age: Age

sibsp: Number of siblings/spouses aboard




7.1

In [17]:
# Remove rows with any missing values
df_cleaned = df.dropna()

We use df.dropna() to ensure that our analysis includes all the data without removing columns but it focuses only on complete rows.

An example could be using it to analyze survival rates based on age, sex, and fare, and there are rows with missing values in age and fare. Using df.dropna() will allow us to analyze the complete data without removing columns.

7.2

In [18]:
# Remove the 'embarked' column due to high missing data
del df['embarked']

We can use del df['col'] when a column has a significant amount of missing data that is crucial for our analysis. This would remove the column entirely. For example, if we wanted to analyze the embarked column, which represents the port of embarkation, we would focus on survival rates and not on embarkation details given the large amount of missing data and so del df['col'] would be the preferred method for a column that is not central to the main goal of the analysis.

7.3

Applying del df['col'] before df.dropna() is important since removing columns with excessive missing data first prevents them from affecting the row removal process. This allows df.dropna() to remove non-essential columns, which improves efficiency and accuracy, while also preventing unnecessary row deletions based on columns that were not valuable or had too many missing values to begin. This then allows the cleaning up of the remaining data, ensuring that only relevant data is used.


7.4

In [20]:
import pandas as pd

# Load the Titanic dataset
df = pd.read_csv('https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv')

# Check initial shape
print("Initial shape:", df.shape)
print(df.info())

Initial shape: (891, 15)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   survived     891 non-null    int64  
 1   pclass       891 non-null    int64  
 2   sex          891 non-null    object 
 3   age          714 non-null    float64
 4   sibsp        891 non-null    int64  
 5   parch        891 non-null    int64  
 6   fare         891 non-null    float64
 7   embarked     889 non-null    object 
 8   class        891 non-null    object 
 9   who          891 non-null    object 
 10  adult_male   891 non-null    bool   
 11  deck         203 non-null    object 
 12  embark_town  889 non-null    object 
 13  alive        891 non-null    object 
 14  alone        891 non-null    bool   
dtypes: bool(2), float64(2), int64(4), object(7)
memory usage: 92.4+ KB
None


In [21]:
# Remove the 'embarked' column
del df['embarked']

# Remove rows with any remaining missing values
df_cleaned = df.dropna()

# Check the shape of the cleaned DataFrame
print("Cleaned shape:", df_cleaned.shape)
print(df_cleaned.info())


Cleaned shape: (182, 14)
<class 'pandas.core.frame.DataFrame'>
Index: 182 entries, 1 to 889
Data columns (total 14 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   survived     182 non-null    int64  
 1   pclass       182 non-null    int64  
 2   sex          182 non-null    object 
 3   age          182 non-null    float64
 4   sibsp        182 non-null    int64  
 5   parch        182 non-null    int64  
 6   fare         182 non-null    float64
 7   class        182 non-null    object 
 8   who          182 non-null    object 
 9   adult_male   182 non-null    bool   
 10  deck         182 non-null    object 
 11  embark_town  182 non-null    object 
 12  alive        182 non-null    object 
 13  alone        182 non-null    bool   
dtypes: bool(2), float64(2), int64(4), object(6)
memory usage: 18.8+ KB
None


Justification

Before: The dataset includes all columns with some having missing data. Assume the embarked column is so.

Initial shape: (714, 7)

After: Dataset is cleaned by removing less important columns with high amounts of missing data and then deleting rows with any remaining missing values. This ensures an analysis based on relevant AND complete dataset.

Cleaned shape: (714, 6)

PART EIGHT: POSTLECTURE

8.1

In [23]:
import pandas as pd

url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

print(df.columns)  # To get the names of the columns
print(df.head())   # To see the first few rows

The groupby function splits the data into groups based on some criteria. Then, the describe() method provides summary statistics for the data in each group:

df.groupby("col1"): Groups the data by the values in col1.

["col2"]: Selects the col2 column from the grouped data.

.describe(): Computes summary statistics for each group in col2.

EXAMPLE WITH THIS DATASET

Let's analyze the age column grouped by the class column. 

In [24]:
group_by_class = df.groupby("class")["age"].describe()
print(group_by_class)


        count       mean        std   min   25%   50%   75%   max
class                                                            
First   186.0  38.233441  14.802856  0.92  27.0  37.0  49.0  80.0
Second  173.0  29.877630  14.001077  0.67  23.0  29.0  36.0  70.0
Third   355.0  25.140620  12.495398  0.42  18.0  24.0  32.0  74.0


The result of df.groupby("class")["age"].describe() provides a summary statistics for the age column, broken down by each class (e.g., First, Second, Third), including metrics like count, mean, standard deviation, minimum, and various percentiles.


8.2

df.describe() provides summary statistics for each numeric column in the DataFrame.

The count value in df.describe() represents the number of non-missing (non-NaN) values for each column across the entire DataFrame. It essentially tells you how many data points are available for each column, excluding any missing values. 

For example, if df has 1000 rows but some columns have missing values, the count of rows (data points) for those columns will be less than 1000

df.groupby("col1")["col2"].describe() provides summary statistics for col2 within each group defined by col1.

The count value here represents the number of non-missing values within each group defined by col1. Each group might have a different number of non-missing values depending on the distribution of missing data in col2 within that group.

For example, (df.groupby("class")["age"].describe()) shows summary statistics for age within each class, including how many non-missing values (count) are present in each class group.

These two functions capture different views of the dataset. While df.describe() provides a high-level view of missing data across the entire dataset, df.groupby("col1")["col2"].describe() provides a more granular view, showing how missing data and summary statistics differ by subgroup. This can be crucial for understanding patterns and discrepancies within different categories of your data.

8.3

After trying out the errors that occur, I think that ChatGPT works better than a Google Search. This is because the chatbot essentially does the same thing as a google search except it removes the task of needing to sift through each search result that appears. The chatbot seems to provide multiple solutions from multiple sources more efficiently and quickly.

In [ ]:
PART NINE: POST LECTURE

Yes, mostly

CHATLOGS:

https://chatgpt.com/share/ad56da47-1cee-489a-97dd-73c83a01c0c5

https://chatgpt.com/share/109ba76a-b6bb-44ae-9ce0-73492c3579a3

https://chatgpt.com/share/9c4c0e6e-0a0a-44c6-bb32-95b334dffb6a

BRIEF SUMMARY:

1. **Difference Between Attributes and Methods**:
   - **Attributes** are properties of an object, accessed directly (e.g., `df.shape`).
   - **Methods** are functions that perform actions on the object and require parentheses to execute (e.g., `df.describe()`).

2. **Functions in Programming vs. Mathematics**:
   - **Programming Functions**: Execute code to perform operations or calculations.
   - **Mathematical Functions**: Define relationships between variables.

3. **Summary Statistics from `df.describe()`**:
   - **Count**: Number of non-missing values.
   - **Mean**: Average value.
   - **Std**: Standard deviation, measure of spread.
   - **Min**: Minimum value.
   - **25%**: 25th percentile.
   - **50%**: Median (50th percentile).
   - **75%**: 75th percentile.
   - **Max**: Maximum value.

4. **Handling Missing Data**:
   - **`df.dropna()`**: Removes rows with any missing values. Useful when you want to retain columns but need complete rows for analysis.
   - **`del df['col']`**: Removes a column, useful when the column has excessive missing data or is irrelevant.
   - **Combining**: Remove columns first to simplify data and then drop rows to clean up the remaining data.

5. **Application to Titanic Dataset**:
   - Demonstrated removing a column with high missing data (`embarked`) and then dropping rows with remaining missing values.
   - The dataset was cleaned to ensure only complete records were used, improving the quality of the analysis.
   
6. **Analyzing Titanic Data**:
   - You started by exploring how to use a ChatBot to troubleshoot code with the Titanic dataset.
   - You learned about the `df.groupby("col1")["col2"].describe()` function, which groups data by a column and then provides summary statistics for another column within each group.

7. **Understanding `df.groupby("col1")["col2"].describe()`**:
   - This function groups data by `col1`, selects `col2`, and then applies `describe()` to get summary statistics for `col2` within each group.
   - Example provided: Analyzing the `age` column grouped by the `class` column.

8. **Handling Missing Values**:
   - **`df.describe()`**: Shows summary statistics for each column in the entire DataFrame, with `count` indicating the number of non-missing values across the dataset.
   - **`df.groupby("col1")["col2"].describe()`**: Shows summary statistics for `col2` within each group defined by `col1`, with `count` indicating non-missing values within each group.

9. **Key Differences**:
   - **Scope of Counting**: `df.describe()` covers the entire DataFrame, while `df.groupby("col1")["col2"].describe()` is specific to each group.
   - **Purpose**: `df.describe()` provides a broad overview, while `df.groupby("col1")["col2"].describe()` offers detailed insights into subgroup-specific patterns.

This summary captures the key points of our discussion on analyzing the Titanic dataset and understanding how different methods handle missing values and provide insights. There was also the additional trouble shooting that was discussed during the process of this homework.